### PR TITLE
feat: create the ability to limit cache

### LINKED ISSUE #482: [REQUEST] Limiting cache
**What is the feature you're proposing?**
As of right now, caching has no limit.
I feel like there should be a way to set a limit for each type of cache, preferably in the `Client` instance.

Something like:
```py
bot = interactions.Client(
    ...,
    cache_limits=interactions.CacheLimits(
        messages=100,
        guilds=50,
        ...
    )
)
```

This is the best thing that I can think of off the top of my head.
Another way is to ~~do what dis-snek did~~ do something with kwargs and have something like this:
```py
bot = interactions.Client(
    ...,
    message_cache=100,
    guild_cache=50,
    ...
)
```

**Additional Information**
- [ ] My feature request is related to an existing Issue.
    - Issue (if referenceable):

### PR DESCRIPTION
**Pull request by Catalyst4222**
## About

This pr adds in the ability to limit cache
Draft until proper default limits are set

## Checklist

- [x] The ``pre-commit`` code linter has been run over all edited files to ensure the code is linted.
- [x] I've ensured the change(s) work on `3.8.6` and higher.

I've made this pull request: (check all that apply)
  - [ ] For the documentation
  - [x] To add a new feature
  - [x] As a general enhancement
  - [ ] As a refactor of the library/the library's code
  - [ ] To fix an existing bug
  - [x] To resolve #482 

This is:
  - [ ] A breaking change

### CODE REVIEW

File: interactions/utils/dict_caches.py [hunk 1]
```diff
@@ -0,0 +1,45 @@
+from collections import OrderedDict
+from typing import Generic, TypeVar
+
+_KT = TypeVar("_KT")
+_VT = TypeVar("_VT")
+
```
Comment (sillyElvy) [inline interactions/utils/dict_caches.py:8]: youre missing __all__ here  (im aware that its draft)

Comment (Catalyst4222) [reply interactions/utils/dict_caches.py:8]: What do you mean by all?

Comment (sillyElvy) [reply interactions/utils/dict_caches.py:8]: `__all__`, forgot that it formats

File: interactions/client/bot.py [hunk 2]
```diff
@@ -75,21 +77,25 @@ class Client:
     def __init__(
         self,
         token: str,
+        cache_limits: Optional[Dict[type, int]] = None,
+        intents: Intents = Intents.DEFAULT,
+        shards: Optional[List[Tuple[int]]] = None,
+        default_scope=None,  # todo typehint
```
Comment (Toricane) [inline interactions/client/bot.py]:
```suggestion
        default_scope: Union[int, Snowflake, List[int, Snowflake]] = None,
```
that would work, right?

**Review by Toricane — CHANGES_REQUESTED** [review]
It looks really good, just one small thing

File: interactions/api/cache.py [hunk 3]
```diff
@@ -38,8 +40,13 @@ class Storage(Generic[_T]):
     def __repr__(self) -> str:
         return f"<{self.__class__.__name__} object containing {len(self.values)} items.>"
 
-    def __init__(self) -> None:
-        self.values: Dict["Key", _T] = {}
+    def __init__(self, limit: Optional[int] = float("inf")) -> None:
+        """
+
+        :param limit: The maximum number of items to store
+        :type limit: Optional[int]
```
Comment (mAxYoLo01) [inline interactions/api/cache.py]:
```suggestion
        :param Optional[int] limit: The maximum number of items to store
```
To account for the new way to write docstrings

File: interactions/client/bot.py [hunk 4]
```diff
@@ -75,21 +77,25 @@ class Client:
...
+        presence: Optional[ClientPresence] = None,
+        _logging: Union[bool, int] = None,
+        disable_sync: bool = False,
         **kwargs,
     ) -> None:
         self._loop: AbstractEventLoop = get_event_loop()
-        self._http: HTTPClient = token
-        self._intents: Intents = kwargs.get("intents", Intents.DEFAULT)
-        self._shards: List[Tuple[int]] = kwargs.get("shards", [])
+        self._http: HTTPClient = token  # noqa
```
Comment (sillyElvy) [inline interactions/client/bot.py]: You could make it Union with string

**Review by sillyElvy — CHANGES_REQUESTED** [review]

File: interactions/client/bot.py (hunk 4 above)
Comment (Catalyst4222) [reply interactions/client/bot.py]: It was left as the httpclient from when you moved it later into the starting process, so I just left it like that. I'll change it as soon as I'm able to

File: interactions/client/bot.py [hunk 5]
```diff
@@ -75,21 +77,25 @@ class Client:
     def __init__(
         self,
         token: str,
+        cache_limits: Optional[Dict[type, int]] = None,
+        intents: Intents = Intents.DEFAULT,
+        shards: Optional[List[Tuple[int]]] = None,
+        default_scope: Union[int, Snowflake, List[int, Snowflake]] = None,
```
Comment (Toricane) [inline interactions/client/bot.py]:
```suggestion
        default_scope: Union[int, Snowflake, List[Union[int, Snowflake]]] = None,
```
Oops

**Review by Toricane — CHANGES_REQUESTED** [review]

File: interactions/api/cache.py [hunk 6]
```diff
@@ -38,8 +40,12 @@ class Storage(Generic[_T]):
     def __repr__(self) -> str:
         return f"<{self.__class__.__name__} object containing {len(self.values)} items.>"
 
-    def __init__(self) -> None:
-        self.values: Dict["Key", _T] = {}
+    def __init__(self, limit: Optional[int] = float("inf")) -> None:
+        """
+
```
Comment (Toricane) [inline interactions/api/cache.py]: ?

File: interactions/client/bot.py [hunk 7]
```diff
@@ -75,21 +77,25 @@ class Client:
     def __init__(
         self,
         token: str,
+        cache_limits: Optional[Dict[type, int]] = None,
+        intents: Intents = Intents.DEFAULT,
+        shards: Optional[List[Tuple[int]]] = None,
+        default_scope: Union[int, Snowflake, List[Union[int, Snowflake]]] = None,
```
Comment (Toricane) [inline interactions/client/bot.py]:
```suggestion
        default_scope: Optional[Union[int, Snowflake, List[Union[int, Snowflake]]]] = None,
```

**Review by Toricane — CHANGES_REQUESTED** [review]

File: interactions/api/cache.py (hunk 6 above)
Comment (Catalyst4222) [reply interactions/api/cache.py]: Just a formatting artifact

File: interactions/api/cache.py [hunk 8]
```diff
@@ -38,8 +40,11 @@ class Storage(Generic[_T]):
     def __repr__(self) -> str:
         return f"<{self.__class__.__name__} object containing {len(self.values)} items.>"
 
-    def __init__(self) -> None:
-        self.values: Dict["Key", _T] = {}
+    def __init__(self, limit: Optional[int] = float("inf")) -> None:
+        """
+        :param Optional[int] limit: The maximum number of items to store
+        """
```
Comment (Toricane) [inline interactions/api/cache.py]:
```suggestion
    def __init__(self, limit: Optional[int] = None) -> None:
        """
        :param Optional[int] limit: The maximum number of items to store
        """
        if not limit:
            limit = float("inf")
```

**Review by Toricane — CHANGES_REQUESTED** [review]

File: interactions/api/cache.py (hunk 8 above)
Comment (Catalyst4222) [reply interactions/api/cache.py:46]: Why is this needed?

**Review by Toricane — APPROVED** [review]
Hmmmmm LGTM

File: interactions/api/cache.py [hunk 9]
```diff
@@ -12,6 +12,8 @@
     overload,
 )
 
+import interactions
```
Comment (sillyElvy) [inline interactions/api/cache.py:15]: why are you doing this?

Comment (Catalyst4222) [reply interactions/api/cache.py:15]: It avoids a circular import for the LRUDict

Comment (sillyElvy) [reply interactions/api/cache.py:15]: Alright

**Review by sillyElvy — APPROVED** [review]
Lgtm ig

**Comment by sillyElvy** [comment]
Merge conflicts :D

**Comment by Catalyst4222** [comment]
F30EE5FE-A567-496C-A1FD-1E4888B2C09E

**Pull request merged**